// proximity-hcsr04/index.js
// HC-SR04 proximity sensor
//
// http://letsmakerobots.com/node/30209

var tessel = require('tessel')
  , EventEmitter = require('events').EventEmitter;

function Proximity (trig, echo) {
  // Module takes in two lines, a trigger and an echo line
  // Echo returns a pulse width after signal is emitted
  this.trig = trig.output();
  this.echo = echo.input();

  // The "record" function traces when the 'rise' and 'fall'
  // of a signal is, firing events and tracing the event times.
  // We use this to track pulse width.
  echo.record();
}

Proximity.prototype = new EventEmitter();

Proximity.prototype.ping = function () {
  // Once we receive the falling edge of a signal, we track
  // the signal width and determine if it's within valid range.
  // Interpret signal to get centimeters.
  var prox = this;
  this.echo.once('fall', function (lastrise, fall) {
    var cm = ((fall - lastrise) / 2) / 29.1;
    if (cm >= 200 || cm <= 0) {
      prox.emit('ping', { cm: cm })
    }
  });

  // Pulse trigger for 50us
  this.trig.pulse(50);
}

// Call to return a Proximity object
function initialize (trig, echo) {
  return new Proximity(trig, echo);
}

exports.initialize = initialize;
